### PR TITLE
Check if stopwatch is started before stopping

### LINKED ISSUE #3272: Stopwatch not started
**Affected version(s)**

Contao 4.9.17

**Description**

Since the update to 4.9.17 I get a lot of this `LogicExceptions`:

```
Event "contao.frontend_module.header (ID 10)" is not started.
Event "contao.frontend_module.newsreader (ID 16)" is not started.
```

Stacktrace:
```
LogicException:
Event "contao.frontend_module.newsreader (ID 16)" is not started.

  at vendor/symfony/stopwatch/Section.php:142
  at Symfony\Component\Stopwatch\Section->stopEvent('contao.frontend_module.newsreader (ID 16)')
     (vendor/symfony/stopwatch/Stopwatch.php:126)
  at Symfony\Component\Stopwatch\Stopwatch->stop('contao.frontend_module.newsreader (ID 16)')
     (vendor/contao/core-bundle/src/Resources/contao/elements/ContentModule.php:103)
  at Contao\ContentModule->generate()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:607)
  at Contao\Controller::getContentElement(object(ContentModel), 'main')
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:218)
  at Contao\ModuleArticle->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:214)
  at Contao\Module->generate()
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:71)
  at Contao\ModuleArticle->generate(false)
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:536)
  at Contao\Controller::getArticle(object(ArticleModel), false, false, 'main')
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:377)
  at Contao\Controller::getFrontendModule('0', 'main')
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:168)
  at Contao\PageRegular->prepare(object(PageModel))
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:48)
  at Contao\PageRegular->getResponse(object(PageModel), true)
     (vendor/contao/core-bundle/src/Resources/contao/controllers/FrontendIndex.php:337)
  at Contao\FrontendIndex->renderPage(object(PageModel))
     (vendor/symfony/http-kernel/HttpKernel.php:158)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:80)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:201)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (public/index.php:43)
```

Possibly related to #2940

### PR DESCRIPTION
**Pull request by bytehead**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #3272
| Docs PR or issue | n/a

This could be a fix for Contao 4.9.20+

### CODE REVIEW

**Review by fritzmg — DISMISSED** [review]

File: core-bundle/src/Resources/contao/elements/ContentModule.php [hunk 1]
```diff
@@ -98,9 +100,9 @@ public function generate()
 
 		$strBuffer = $objModule->generate();
 
-		if (isset($objStopwatch))
+		if (isset($objStopwatch) && $objStopwatch->isStarted($strStopWatchId))
```
Comment (m-vo) [inline core-bundle/src/Resources/contao/elements/ContentModule.php:103]: Did you find out why the stopwatch isn't started at this point? Is something during `$objModule->generate();` stopping it?

Comment (bytehead) [reply core-bundle/src/Resources/contao/elements/ContentModule.php:103]: No, but as I understand it: different requests do have different stopwatch sections and there we get the little mess: https://github.com/symfony/symfony/issues/36623

Comment (bytehead) [reply core-bundle/src/Resources/contao/elements/ContentModule.php:103]: In my real life demo all stopwatches get started, but after the first subrequest all following `$stopwatch->stop()` will fail and `objStopwatch->isStarted($strStopWatchId)` will return `false`.

Comment (bytehead) [reply core-bundle/src/Resources/contao/elements/ContentModule.php:103]: I can follow this: https://github.com/symfony/symfony/issues/36762#issuecomment-626352054

Comment (m-vo) [reply core-bundle/src/Resources/contao/elements/ContentModule.php:103]: Ah yes, missed the linked ticket. Thanks!

**Review by m-vo — DISMISSED** [review]

**Comment by fritzmg** [comment]
Btw. I think this should get merged regardless of the original issue. It makes sense anyway to always check whether the stopwatch has actually started before trying to stop it. Similar to always checking whether a session is present and started etc. 
Also the stop watch could have been stopped by a third party in between.

**Comment by bytehead** [comment]
> Also the stop watch could be stopped by a third party in between.

Yes. 

Symfony itself checks in [their code using the stopwatch](https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/Messenger/Middleware/TraceableMiddleware.php#L74)  if it's started upfront.

**Review by Toflar — DISMISSED** [review]

**Review by aschempp — DISMISSED** [review]
I agree we might need to check if a stopwatch was started, but I dislike the fact that it would be silently ignored. If we do start a stopwatch, shouldn't we be the only ones allowed to stop it? Otherwise the stopwatch will not give useful results and no-one will notice 🤷

**Comment by bytehead** [comment]
If we check if it's started upfront, no annoying exception will be thrown in case of an error (the stopwatch is just for debugging purposes anyway). You would also notice the error introduced from Symfony in the profiler, as the timing information will not show up for these modules/content elements which are rendered as fragments. So first people don't get annoying errors and you would notice it if something is wrong when you check the performance tab in the profiler and search for the data you want.

**Comment by bytehead** [comment]
Status quo is now unusable in dev mode (and the information is not _more correctly_ than if it was checked upfront and not shown) and that's a problem 🙃

**Comment by aschempp** [comment]
> You would also notice the error introduced from Symfony in the profiler, as the timing information will not show up for these modules/content elements which are rendered as fragments.

Is that true? Isn't the stopwatch just stopped incorrectly at another point (on kernel.response in the current issue)?

**Comment by bytehead** [comment]
Only if somebody stopps it. Otherwise it gets omitted (without an error anymore in this case).

**Comment by aschempp** [comment]
> Only if somebody stopps it. Otherwise it gets omitted (without an error anymore in this case).

No idea why that should be possible with a stopwatch, but apparently I don't know enough about how it works 😂
I trust you on this 👍

**Review by aschempp — APPROVED** [review]

**Review by Toflar — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @bytehead.

**Pull request merged**